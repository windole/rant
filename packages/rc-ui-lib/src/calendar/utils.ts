/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
