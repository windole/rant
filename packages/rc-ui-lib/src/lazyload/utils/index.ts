import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
