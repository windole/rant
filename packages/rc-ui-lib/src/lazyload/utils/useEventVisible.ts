import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
