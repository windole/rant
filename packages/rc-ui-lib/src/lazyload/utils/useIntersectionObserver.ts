import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
