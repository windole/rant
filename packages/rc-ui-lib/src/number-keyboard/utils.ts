import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
