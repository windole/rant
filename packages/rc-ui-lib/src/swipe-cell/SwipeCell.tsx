/* eslint-disable @typescript-eslint/no-shadow */
import React, { forwardRef, useContext, useRef, useImperativeHandle, RefObject } from 'react';
import classNames from 'classnames';
import { useSpring, animated } from '@react-spring/web';
import { useDrag } from '@use-gesture/react';

import useClickAway from '../hooks/use-click-away';
import useMountedRef from '../hooks/use-mounted-ref';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';
import { callInterceptor } from '../utils/interceptor';
import { nearest } from '../utils/nearest';
import { SwipeCellInstance, SwipeCellPosition, SwipeCellProps, SwipeCellSide } from '.';
import { isDef } from '../utils';

const SwipeCell = forwardRef<SwipeCellInstance, SwipeCellProps>((props, ref) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('swipe-cell', prefixCls);

  const rootRef = useRef<HTMLDivElement>(null);
  const leftRef = useRef<HTMLDivElement>();
  const rightRef = useRef<HTMLDivElement>();

  const opened = useRef(false);
  const lockClick = useRef(false);
  const draggingRef = useRef(false);

  const mountedRef = useMountedRef();

  const getWidth = (elementRef: RefObject<HTMLDivElement>) =>
    elementRef.current ? elementRef.current.offsetWidth : 0;

  function getLeftWidth() {
    return isDef(props.leftWidth) ? +props.leftWidth : getWidth(leftRef);
  }
  function getRightWidth() {
    return isDef(props.rightWidth) ? +props.rightWidth : getWidth(rightRef);
  }

  const [{ x }, api] = useSpring(
    () => ({
      x: 0,
      config: {
        tension: 200,
        friction: 30,
      },
    }),
    [],
  );

  const bind = useDrag(
    (state) => {
      if (!props.disabled) {
        lockClick.current = true;
        draggingRef.current = true;
        const [offsetX] = state.offset;
        if (state.last) {
          const leftWidth = getLeftWidth();
          const rightWidth = getRightWidth();
          let position = offsetX + state.velocity[0] * state.direction[0] * 50;
          if (offsetX > 0) {
            position = Math.max(0, position);
          } else {
            position = Math.min(0, position);
          }
          const nearestPosition = nearest([-rightWidth, 0, leftWidth], position);
          api.start({
            x: nearestPosition,
          });
          opened.current = nearestPosition !== 0;
          window.setTimeout(() => {
            if (mountedRef.current) {
              draggingRef.current = false;
              lockClick.current = false;
            }
          });
        } else {
          api.start({
            x: offsetX,
            immediate: true,
          });
        }
      }
    },
    {
      from: () => [x.get(), 0],
      bounds: () => {
        const leftWidth = getLeftWidth();
        const rightWidth = getRightWidth();
        return {
          left: -rightWidth,
          right: leftWidth,
        };
      },
      axis: 'x',
      preventScroll: true,
      pointer: { touch: true },
    },
  );

  const open = (side: SwipeCellSide) => {
    opened.current = true;
    const offset = side === 'left' ? getLeftWidth() : -getRightWidth();

    api.start({
      x: offset,
    });

    props.onOpen?.({
      position: side,
      name: props.name,
    });
  };

  const close = (position?: SwipeCellPosition) => {
    api.start({
      x: 0,
    });
    if (opened.current) {
      opened.current = false;
      props.onClose?.({
        position,
        name: props.name,
      });
    }
  };

  const onClick = (position: SwipeCellPosition = 'outside') => {
    props.onClick?.(position);
    if (opened.current && !lockClick.current) {
      callInterceptor({
        interceptor: props.beforeClose,
        args: [
          {
            name: props.name,
            position,
            instance: {
              close: () => close(position),
            },
          },
        ],
        done: () => close(position),
      });
    }
  };

  const getClickHandler = (position: SwipeCellPosition, stop?: boolean) => (event) => {
    if (stop) {
      event.preventDefault();
      event.stopPropagation();
    }
    onClick(position);
  };

  const renderSideContent = (
    side: SwipeCellSide,
    sideRef: React.MutableRefObject<HTMLDivElement>,
  ) => {
    const contentSlot = props[side];
    return contentSlot ? (
      <div ref={sideRef} className={classNames(bem(side))} onClick={getClickHandler(side, true)}>
        {contentSlot}
      </div>
    ) : null;
  };

  useImperativeHandle(ref, () => ({
    open: (side: SwipeCellSide) => open(side),
    close,
  }));

  useClickAway(rootRef, () => onClick('outside'), 'touchstart', props.hideOnClickOutside);

  return (
    <div
      ref={rootRef}
      {...bind()}
      className={classNames(bem())}
      onClickCapture={(e) => {
        if (draggingRef.current) {
          e.stopPropagation();
          e.preventDefault();
        }
      }}
      onClick={getClickHandler('cell')}
    >
      <animated.div className={classNames(bem('wrapper'))} style={{ x }}>
        {renderSideContent('left', leftRef)}
        <div
          className={classNames(bem('content'))}
          onClickCapture={(e) => {
            if (x.goal !== 0) {
              e.preventDefault();
              e.stopPropagation();
              api.start({
                x: 0,
              });
            }
          }}
        >
          <animated.div
            style={{
              pointerEvents: x.to((v) => (v !== 0 && x.goal !== 0 ? 'none' : 'unset')),
            }}
          >
            {props.children}
          </animated.div>
        </div>
        {renderSideContent('right', rightRef)}
      </animated.div>
    </div>
  );
});

SwipeCell.defaultProps = {
  disabled: false,
  hideOnClickOutside: true,
};

export default SwipeCell;
