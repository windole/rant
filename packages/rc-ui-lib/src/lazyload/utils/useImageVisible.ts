import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
