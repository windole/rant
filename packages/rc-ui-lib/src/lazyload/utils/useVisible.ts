import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
