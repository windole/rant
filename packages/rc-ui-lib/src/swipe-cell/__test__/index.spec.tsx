/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, cleanup, waitFor, act, fireEvent } from '@testing-library/react';
import TestsEvent from '../../../tests/events';
import { sleep } from '../../../tests/utils';
import { Button, Cell } from '../..';
import SwipeCell, { SwipeCellInstance, SwipeCellProps, SwipeCellSide } from '..';

interface ElementRect {
  bottom: number;
  height: number;
  left: number;
  right: number;
  top: number;
  width: number;
  x: number;
  y: number;
}

const swipeStyle = {
  width: '100px',
  height: '100px',
};

const $props = {
  leftWidth: 100,
  rightWidth: 100,
};

describe('SwipeCell test with testing library', () => {
  let eachClears: Array<() => void> = [];

  const mockGetBoundingClientRect = (rect: Partial<ElementRect>): (() => void) => {
    const originMethod = Element.prototype.getBoundingClientRect;

    Element.prototype.getBoundingClientRect = jest.fn(
      () =>
        ({
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          width: 0,
          height: 0,
          ...rect,
        } as any),
    );

    return () => {
      Element.prototype.getBoundingClientRect = originMethod;
    };
  };

  const mockOffset = (c: Element) => {
    Object.defineProperty(c, 'offsetWidth', {
      configurable: true,
      get: () => 100,
    });
  };

  function createSwipeCell(props?: Partial<SwipeCellProps>) {
    const swipeCellRef = React.createRef<SwipeCellInstance>();

    const { queryByTestId, container, rerender, debug } = render(
      <SwipeCell
        ref={swipeCellRef}
        style={swipeStyle}
        name="swipeCell"
        left={<Button square type="primary" text="选择" />}
        right={<Button square type="danger" text="删除" />}
        {...props}
      >
        <Cell title="单元格" value="内容" />
      </SwipeCell>,
    );

    return {
      container,
      rerender,
      swipeCellRef,
      debug,
      queryByTestId,
    };
  }

  beforeEach(() => {
    jest.useFakeTimers();

    const clear = mockGetBoundingClientRect({
      width: 100,
      height: 100,
    });

    eachClears.push(clear);
  });

  afterEach(() => {
    eachClears.forEach((clear) => clear());
    eachClears = [];

    cleanup();
    jest.restoreAllMocks();
  });

  it('should swipe to specific side after calling the open method', async () => {
    const onOpen = jest.fn();

    const { swipeCellRef, container } = createSwipeCell({
      ...$props,
      onOpen,
    });

    const open = (side: SwipeCellSide) => {
      swipeCellRef.current.open(side);
    };

    await waitFor(async () => {
      open('left');
    });

    await sleep(1000);

    expect(onOpen).toHaveBeenCalledWith({
      name: 'swipeCell',
      position: 'left',
    });

    expect(container).toMatchSnapshot();
  });

  it('should emit onClick event when click swipe-cell', async () => {
    const onClick = jest.fn();
    const { container } = createSwipeCell({
      ...$props,
      onClick,
    });

    const wrap = container.querySelector('.rc-swipe-cell');
    await fireEvent.click(wrap);

    expect(onClick).toHaveBeenCalledWith('cell');
  });

  it('should close after calling close method', async () => {
    const onClose = jest.fn();

    const { swipeCellRef } = createSwipeCell({
      ...$props,
      onClose,
    });

    await waitFor(async () => {
      swipeCellRef.current.open('left');
    });

    await waitFor(async () => {
      swipeCellRef.current.close('outside');
    });

    await sleep(1000);

    expect(onClose).toHaveBeenCalledWith({
      name: 'swipeCell',
      position: 'outside',
    });
  });

  it('should allow to drag to show left part', async () => {
    const { container } = createSwipeCell({
      ...$props,
    });

    const track = container.querySelector('.rc-swipe-cell__wrapper');
    const leftSide = container.querySelector('.rc-swipe-cell__left');

    mockOffset(leftSide);
    await TestsEvent.triggerDrag(track, [100, 0]);

    await sleep(1000);

    expect(container).toMatchSnapshot();
  });

  it('should allow to drag to show right part', async () => {
    const { container } = createSwipeCell({
      ...$props,
    });

    const track = container.querySelector('.rc-swipe-cell__wrapper');
    const rightSide = container.querySelector('.rc-swipe-cell__right');

    mockOffset(rightSide);
    await TestsEvent.triggerDrag(track, [-100, 0]);

    await sleep(1000);

    expect(container).toMatchSnapshot();
  });

  it('should call beforeClose before closing', async () => {
    let position;
    let instance;
    const { container, swipeCellRef } = createSwipeCell({
      ...$props,
      // eslint-disable-next-line consistent-return
      beforeClose(params) {
        ({ position, instance } = params);
        if (position === 'right') {
          instance.close();
        } else {
          return true;
        }
      },
    });

    const track = container.querySelector('.rc-swipe-cell__wrapper');
    const rightSide = container.querySelector('.rc-swipe-cell__right');
    const leftSide = container.querySelector('.rc-swipe-cell__left');

    mockOffset(rightSide);

    await waitFor(async () => {
      swipeCellRef.current.open('right');
    });

    fireEvent.click(track);
    expect(position).toEqual('cell');

    await waitFor(async () => {
      swipeCellRef.current.open('right');
    });

    fireEvent.click(rightSide);
    expect(position).toEqual('right');

    await waitFor(async () => {
      swipeCellRef.current.open('left');
    });

    fireEvent.click(leftSide);
    expect(position).toEqual('left');
  });

  it('should reset transform after short dragging', async () => {
    const { container } = createSwipeCell();

    const track = container.querySelector('.rc-swipe-cell__wrapper');
    const leftSide = container.querySelector('.rc-swipe-cell__left');

    mockOffset(leftSide);

    await TestsEvent.triggerDrag(track, [-5, 0]);

    act(() => {
      jest.advanceTimersByTime(500);
    });

    expect(container).toMatchSnapshot();
  });

  it('should not allow to drag when using disabled prop', async () => {
    const { container } = createSwipeCell({
      ...$props,
      disabled: true,
    });

    const track = container.querySelector('.rc-swipe-cell__wrapper');
    const leftSide = container.querySelector('.rc-swipe-cell__left');

    mockOffset(leftSide);
    await TestsEvent.triggerDrag(track, [100, 0]);

    await sleep(1000);

    expect(container).toMatchSnapshot();
  });

  it('should render correct when click content', async () => {
    const { container, swipeCellRef } = createSwipeCell({
      ...$props,
    });

    const content = container.querySelector('.rc-swipe-cell__content');

    await waitFor(async () => {
      swipeCellRef.current.open('left');
    });

    await fireEvent.click(content);

    await sleep(1000);

    expect(container).toMatchSnapshot();
  });
});
