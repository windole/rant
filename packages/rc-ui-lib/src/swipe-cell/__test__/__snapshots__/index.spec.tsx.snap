// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeCell test with testing library should allow to drag to show left part 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: translate3d(100px,0,0);"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: none;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeCell test with testing library should allow to drag to show right part 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: translate3d(-100px,0,0);"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: none;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeCell test with testing library should not allow to drag when using disabled prop 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: none;"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: unset;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeCell test with testing library should render correct when click content 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: none;"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: unset;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeCell test with testing library should reset transform after short dragging 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: none;"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: unset;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeCell test with testing library should swipe to specific side after calling the open method 1`] = `
<div>
  <div
    class="rc-swipe-cell"
  >
    <div
      class="rc-swipe-cell__wrapper"
      style="transform: translate3d(100px,0,0);"
    >
      <div
        class="rc-swipe-cell__left"
      >
        <button
          class="rc-button rc-button--primary rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              选择
            </span>
          </div>
        </button>
      </div>
      <div
        class="rc-swipe-cell__content"
      >
        <div
          style="pointer-events: none;"
        >
          <div
            class="rc-cell"
          >
            <div
              class="rc-cell__title"
            >
              单元格
            </div>
            <div
              class="rc-cell__value"
            >
              <span>
                内容
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rc-swipe-cell__right"
      >
        <button
          class="rc-button rc-button--danger rc-button--normal rc-button--square"
        >
          <div
            class="rc-button__content"
          >
            <span
              class="rc-button__text"
            >
              删除
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
